alf/config_util_test.py

```diff
@@ -182,6 +182,63 @@ def sole_init_test_env(x):
         self.assertEqual(alf.get_config_value("sole_init_test_twice.x"), 1)
         self.assertEqual(alf.get_config_value("sole_init_test_env.x"), 1)
 
+        # Test override_config doesn't doesn't overwrite for immutable values.
+        @alf.configurable
+        def override_on_immutable(x):
+            pass
+
+        alf.config("override_on_immutable", x=0, mutable=False)
+        alf.override_config("override_on_immutable", x=1)
+        self.assertEqual(alf.get_config_value("override_on_immutable.x"), 0)
+
+        # Test override_config doesn't doesn't overwrite for immutable values.
+        @alf.configurable
+        def override_on_immutable_and_sole_init(x):
+            pass
+
+        alf.config(
+            "override_on_immutable_and_sole_init",
+            x=0,
+            sole_init=True,
+            mutable=False)
+        alf.override_config("override_on_immutable_and_sole_init", x=1)
+        self.assertEqual(
+            alf.get_config_value("override_on_immutable_and_sole_init.x"), 0)
+
+        # Test that pre_config calls work before a sole_init config call
+        @alf.configurable
+        def pre_config_before(x):
+            pass
+
+        alf.pre_config({"pre_config_before.x": 0})
+        alf.config("pre_config_before", x=1)
+        alf.override_config("pre_config_before", x=1)
+        # sole_init starts to take effect for all calls AFTER the first call.
+        alf.config("pre_config_before", x=1, sole_init=True)
+        with self.assertRaises(RuntimeError) as context:
+            alf.config("pre_config_before", x=2)
+        # If truly immutable, the value should never have been changed
+        self.assertEqual(alf.get_config_value("pre_config_before.x"), 0)
+
+        # Test that pre_config calls after a sole_init config call raises an error
+        @alf.configurable
+        def pre_config_after(x):
+            pass
+
+        alf.config("pre_config_after", x=1, sole_init=True)
+        with self.assertRaises(RuntimeError) as context:
+            alf.pre_config({"pre_config_after.x": 0})
+
+        # Test that calling override_config doesn't affect previous sole_init calls.
+        @alf.configurable
+        def override_no_affect_sole_init(x):
+            pass
+
+        alf.config("override_no_affect_sole_init", x=1, sole_init=True)
+        alf.override_config("override_no_affect_sole_init", x=2)
```

Comment:
should this also throw error?

Reply:
It shouldn't throw an error since `override_config` allows us to set a config's value despite an earlier `sole_init=True` call.

Reply:
Oh, I see.  Maybe rename the function to alf.override_sole_config()?  Just as how you renamed override to override_sole_init.

Reply:
Good idea. Done.